File: web/bench_browser.py
#!/usr/bin/env python3
"""Benchmark one JSONL puzzle at a time; preparation is outside the search cap."""
import argparse
import asyncio
import json
import time
from pathlib import Path

from playwright.async_api import async_playwright
from test_browser import open_client, serve, verify_solution


async def main(args):
    puzzles = [json.loads(line) for line in args.puzzles.read_text().splitlines() if line.strip()]
    with serve() as url, args.output.open('w') as output:
        async with async_playwright() as playwright:
            browser = await getattr(playwright, args.browser).launch()
            page, info = await open_client(browser, url, args.threads)
            for index, puzzle in enumerate(puzzles):
                start = time.monotonic()
                record = {'index': index, 'level': puzzle['level'], 'browser': args.browser,
                          'version': browser.version, **info}
                if 'seed' in puzzle:
                    record['seed'] = puzzle['seed']
                try:
                    result = await asyncio.wait_for(page.evaluate(
                        '(p) => client.solve(p.puzzle, p.budget)',
                        {'puzzle': puzzle, 'budget': args.budget * 1000}), args.budget + 180)
                    if result['solved']:
                        verify_solution(puzzle, result)
                    record.update(result)
                except Exception as error:
                    record['error'] = str(error)
                    await page.close()
                    page, info = await open_client(browser, url, args.threads)
                record['wall_ms'] = (time.monotonic() - start) * 1000
                output.write(json.dumps(record) + '\n')
                output.flush()
                print(json.dumps({k: v for k, v in record.items() if k != 'placements'}), flush=True)
            await browser.close()


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('puzzles', type=Path)
    parser.add_argument('--output', type=Path, required=True)
    parser.add_argument('--threads', type=int, help='Default: all browser-reported CPUs')
    parser.add_argument('--budget', type=int, default=120)
    parser.add_argument('--browser', choices=['chromium', 'firefox', 'webkit'], default='chromium')
    asyncio.run(main(parser.parse_args()))

File: web/bench-browser.mjs
import assert from 'node:assert/strict';
import { open } from 'node:fs/promises';
import { parseArgs } from 'node:util';
import * as playwright from 'playwright';
import { serve } from './serve.mjs';
import { openClient, readPuzzles, stringify, verifySolution } from './browser-tools.mjs';

const { values, positionals } = parseArgs({ allowPositionals: true, options: {
    output: { type: 'string' }, threads: { type: 'string' }, budget: { type: 'string', default: '120' },
    browser: { type: 'string', default: 'chromium' }, 'executable-path': { type: 'string' },
    help: { type: 'boolean', short: 'h' },
} });
if (values.help) {
    console.log('Usage: npm run bench:browser -- PUZZLES.jsonl --output RESULTS.jsonl ' +
        '[--threads N] [--budget 120] [--browser chromium|firefox|webkit] [--executable-path PATH]');
} else {
    assert(positionals.length === 1 && values.output, 'Provide a puzzle file and --output');
    assert(['chromium', 'firefox', 'webkit'].includes(values.browser), 'Invalid browser');
    const budget = Number(values.budget);
    const threads = values.threads === undefined ? null : Number(values.threads);
    assert(Number.isFinite(budget) && budget > 0, 'Invalid budget');
    assert(threads === null || (Number.isInteger(threads) && threads > 0), 'Invalid threads');
    const puzzles = readPuzzles(positionals[0]);
    const server = await serve();
    let browser, output;
    try {
        browser = await playwright[values.browser].launch({ executablePath: values['executable-path'] });
        output = await open(values.output, 'w');
        let { page, info } = await openClient(browser, server.url, threads);
        for (const [index, puzzle] of puzzles.entries()) {
            const start = performance.now();
            const { seed, ...game } = puzzle;
            const record = { index, level: puzzle.level, browser: values.browser,
                version: browser.version(), ...info, ...(seed === undefined ? {} : { seed }) };
            let watchdog;
            try {
                const result = await Promise.race([
                    page.evaluate(({ puzzle, budget }) => client.solve(puzzle, budget), { puzzle: game, budget: budget * 1000 }),
                    new Promise((_, reject) => { watchdog = setTimeout(() => reject(new Error('Browser solve watchdog expired')), (budget + 180) * 1000); }),
                ]);
                if (result.solved) verifySolution(puzzle, result);
                Object.assign(record, result);
            } catch (error) {
                record.error = String(error);
                await page.close();
                ({ page, info } = await openClient(browser, server.url, threads));
            } finally {
                clearTimeout(watchdog);
            }
            record.wall_ms = performance.now() - start;
            await output.write(stringify(record) + '\n');
            const { placements, ...summary } = record;
            console.log(stringify(summary));
        }
    } finally {
        await output?.close();
        await browser?.close();
        await server.close();
    }
}
